sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java

```diff
@@ -689,5 +788,155 @@ public void processElement(ProcessContext c) throws Exception {
     }
   }
 
+  /**
+   * Serializes mutations to ((table name, serialized key), serialized value) tuple.
+   */
+  private static class SerializeMutationsFn
+      extends DoFn<MutationGroup, KV<KV<String, byte[]>, byte[]>> {
+
+    final PCollectionView<SpannerSchema> schemaView;
+
+    private SerializeMutationsFn(PCollectionView<SpannerSchema> schemaView) {
+      this.schemaView = schemaView;
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext c) {
+      MutationGroup g = c.element();
+      Mutation m = g.primary();
+      SpannerSchema schema = c.sideInput(schemaView);
+      String table = m.getTable();
+      MutationGroupEncoder mutationGroupEncoder = new MutationGroupEncoder(schema);
+      byte[] key = new byte[] {};
+      if (m.getOperation() != Mutation.Op.DELETE) {
+        key = mutationGroupEncoder.encodeKey(m);
+      } else if (isPointDelete(m)) {
+        Key next = m.getKeySet().getKeys().iterator().next();
+        key = mutationGroupEncoder.encodeKey(m.getTable(), next);
+      }
+      byte[] value = mutationGroupEncoder.encode(g);
+      c.output(KV.of(KV.of(table, key), value));
+    }
+  }
+
+  private static boolean isPointDelete(Mutation m) {
+    return Iterables.isEmpty(m.getKeySet().getRanges())
+        && Iterables.size(m.getKeySet().getKeys()) == 1;
+  }
+
+  /**
+   * Assigns a partition to the mutation group token based on the sampled data.
+   */
+  private static class AssignPartitionFn
+      extends DoFn<KV<KV<String, byte[]>, byte[]>, KV<String, KV<KV<String, byte[]>, byte[]>>> {
```

Comment:
That's a crazy type signature :)

Reply:
I concur and suggest to add a new class (with a coder) to encapsulate `KV<KV<String, byte[]>, byte[]>`